File: core/src/test/java/org/jetbrains/academy/test/system/core/TestJavaClass.java
package org.jetbrains.academy.test.system.core;

import org.jetbrains.academy.test.system.core.models.Visibility;
import org.jetbrains.academy.test.system.core.models.classes.ClassType;
import org.jetbrains.academy.test.system.core.models.classes.TestClass;
import org.jetbrains.academy.test.system.core.models.method.TestMethod;
import org.jetbrains.academy.test.system.core.models.variable.TestVariable;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import java.util.List;

import static java.util.Collections.emptyList;

public class TestJavaClass {
    private static TestClass javaClassTestClass;

    @BeforeAll
    static void beforeAll() {
        javaClassTestClass = new TestClass(
                "JavaClass",
                "org.jetbrains.academy.test.system.core.testData.java",
                Visibility.PUBLIC,
                ClassType.CLASS,
                List.of(
                        new TestVariable(
                                "PUBLIC_CONSTANT",
                                "int",
                                "1",
                                Visibility.PUBLIC,
                                true,
                                false,
                                true
                        ),
                        new TestVariable(
                                "PRIVATE_CONSTANT",
                                "int",
                                "2",
                                Visibility.PRIVATE,
                                true,
                                false,
                                true
                        ),
                        new TestVariable(
                                "publicStaticVar",
                                "int",
                                "3",
                                Visibility.PUBLIC,
                                false,
                                false,
                                true
                        ),
                        new TestVariable(
                                "privateStaticVar",
                                "int",
                                "4",
                                Visibility.PRIVATE,
                                false,
                                false,
                                true
                        ),
                        new TestVariable(
                                "publicVar",
                                "String",
                                "publicVar",
                                Visibility.PUBLIC,
                                false,
                                false,
                                false
                        ),
                        new TestVariable(
                                "privateVar",
                                "String",
                                "privateVar",
                                Visibility.PRIVATE,
                                false,
                                false,
                                false
                        ),
                        new TestVariable(
                                "customClass",
                                "CustomJavaClass",
                                null,
                                Visibility.PUBLIC,
                                false,
                                false,
                                false
                        ),
                        new TestVariable(
                                "primaryConstructorVar",
                                "String",
                                null,
                                Visibility.PUBLIC,
                                false,
                                true,
                                false
                        )
                ),
                List.of(
                        new TestMethod(
                                "publicMethod",
                                "void",
                                emptyList(),
                                Visibility.PUBLIC
                        ),
                        new TestMethod(
                                "privateMethod",
                                "void",
                                emptyList(),
                                Visibility.PRIVATE
                        ),
                        new TestMethod(
                                "calculateSum",
                                "double",
                                List.of(
                                        new TestVariable("a", "double"),
                                        new TestVariable("b", "double")
                                ),
                                Visibility.PUBLIC
                        ),
                        new TestMethod(
                                "getString",
                                "String",
                                List.of(new TestVariable("string", "String")),
                                Visibility.PRIVATE
                        ),
                        new TestMethod(
                                "processList",
                                "List",
                                List.of(new TestVariable("list", "List")),
                                Visibility.PUBLIC
                        )
                ),
                false,
                emptyList(),
                emptyList()
        );
    }

    @Test
    public void javaClassTest() {
        Class<?> clazz = javaClassTestClass.checkBaseDefinition();
        javaClassTestClass.checkFieldsDefinition(clazz, true);
        javaClassTestClass.checkDeclaredMethods(clazz);
    }
}

File: core/src/test/kotlin/org/jetbrains/academy/test/system/core/testData/java/TestJavaClass.kt
package org.jetbrains.academy.test.system.core.testData.java

import org.jetbrains.academy.test.system.core.models.Visibility
import org.jetbrains.academy.test.system.core.models.classes.TestClass
import org.jetbrains.academy.test.system.core.models.method.TestMethod
import org.jetbrains.academy.test.system.core.models.variable.TestVariable
import org.jetbrains.academy.test.system.core.models.variable.VariableMutability

val javaClassTestClass = TestClass(
    "JavaClass",
    "org.jetbrains.academy.test.system.core.testData.java",
    declaredFields = listOf(
        TestVariable(
            name = "PUBLIC_CONSTANT",
            javaType = "int",
            value = "1",
            visibility = Visibility.PUBLIC,
            mutability = VariableMutability.VAL,
            isInPrimaryConstructor = false,
            isConst = true
        ),
        TestVariable(
            name = "PRIVATE_CONSTANT",
            javaType = "int",
            value = "2",
            visibility = Visibility.PRIVATE,
            mutability = VariableMutability.VAL,
            isInPrimaryConstructor = false,
            isConst = true
        ),
        TestVariable(
            name = "publicStaticVar",
            javaType = "int",
            value = "3",
            visibility = Visibility.PUBLIC,
            mutability = VariableMutability.VAR,
            isInPrimaryConstructor = false,
            isStatic = true
        ),
        TestVariable(
            name = "privateStaticVar",
            javaType = "int",
            value = "4",
            visibility = Visibility.PRIVATE,
            mutability = VariableMutability.VAR,
            isInPrimaryConstructor = false,
            isStatic = true
        ),
        TestVariable(
            name = "publicVar",
            javaType = "String",
            value = "publicVar",
            visibility = Visibility.PUBLIC,
            mutability = VariableMutability.VAR,
            isInPrimaryConstructor = false,
        ),
        TestVariable(
            name = "privateVar",
            javaType = "String",
            value = "privateVar",
            visibility = Visibility.PRIVATE,
            mutability = VariableMutability.VAR,
            isInPrimaryConstructor = false,
        ),
        TestVariable(
            name = "customClass",
            javaType = "CustomJavaClass",
            visibility = Visibility.PUBLIC,
            mutability = VariableMutability.VAR,
            isInPrimaryConstructor = false,
        ),
        TestVariable(
            name = "primaryConstructorVar",
            javaType = "String",
            visibility = Visibility.PUBLIC,
            mutability = VariableMutability.VAR,
            isInPrimaryConstructor = true,
        ),
    ),
    customMethods = listOf(
        TestMethod(
            name = "publicMethod",
            returnTypeJava = "void",
            arguments = emptyList(),
            visibility = Visibility.PUBLIC
        ),
        TestMethod(
            name = "privateMethod",
            returnTypeJava = "void",
            arguments = emptyList(),
            visibility = Visibility.PRIVATE
        ),
        TestMethod(
            name = "calculateSum",
            returnTypeJava = "double",
            arguments = listOf(
                TestVariable(
                    name = "a",
                    javaType = "double",
                ),
                TestVariable(
                    name = "b",
                    javaType = "double",
                ),
            ),
            visibility = Visibility.PUBLIC
        ),
        TestMethod(
            name = "getString",
            returnTypeJava = "String",
            arguments = listOf(
                TestVariable(
                    name = "string",
                    javaType = "String",
                ),
            ),
            visibility = Visibility.PRIVATE
        ),
        TestMethod(
            name = "processList",
            returnTypeJava = "List",
            arguments = listOf(
                TestVariable(
                    name = "list",
                    javaType = "List"
                ),
            ),
            visibility = Visibility.PUBLIC
        ),
    )
)
